File: src/agent_nexus/models/permission.py
"""Permission models: PermissionMode, PermissionConfig, PathRule, PermissionDecision."""

from __future__ import annotations

from enum import StrEnum

from pydantic import Field

from agent_nexus.models._common import FrozenModel


class PermissionMode(StrEnum):
    """Permission modes controlling how Agents interact with the system.

    - DEFAULT: Mutable operations require user confirmation.
    - PLAN: Read-only; all modifications are blocked.
    - FULL_AUTO: No confirmation needed; used in CI/CD and test environments.
    """

    DEFAULT = "default"
    PLAN = "plan"
    FULL_AUTO = "full_auto"


class PathAccess(StrEnum):
    """Access level for a path rule."""

    READ = "read"
    WRITE = "write"
    READ_WRITE = "read-write"
    DENY = "deny"


class PathRule(FrozenModel):
    """A glob-pattern path access rule.

    Used in PermissionConfig to control which file paths an Agent can access.

    Example:
        - pattern: "*.docx"
          access: read-write
        - pattern: "*.env"
          access: deny
    """

    pattern: str = Field(min_length=1)
    access: PathAccess = PathAccess.READ


class PermissionConfig(FrozenModel):
    """Permission configuration for an Agent.

    Maps to the `permissions` section in agent-manifest.yaml.

    Evaluation order (later steps can override earlier ones):
    1. Built-in sensitive path protection (SSH, AWS, GCP, etc.) -- cannot override
    2. denied_tools (blacklist)
    3. allowed_tools (whitelist)
    4. path_rules (glob pattern access control)
    5. denied_commands (dangerous shell command patterns)
    6. mode (DEFAULT/PLAN/FULL_AUTO baseline)
    7. Read-only tool exemption (file_read, grep, etc. always allowed unless FULL_AUTO)

    Example agent-manifest.yaml:
        permissions:
          mode: default
          allowed_tools: [file_read, file_write, mcp__docx__*]
          denied_tools: [bash]
          path_rules:
            - pattern: "*.docx"
              access: read-write
            - pattern: "*.env"
              access: deny
    """

    mode: PermissionMode = PermissionMode.DEFAULT
    allowed_tools: list[str] = Field(default_factory=list)
    denied_tools: list[str] = Field(default_factory=list)
    path_rules: list[PathRule] = Field(default_factory=list)
    denied_commands: list[str] = Field(default_factory=list)


class PermissionDecision(FrozenModel):
    """Result of a permission check.

    Attributes:
        allowed: Whether the action is permitted.
        reason: Human-readable explanation (especially for denials).
        requires_confirmation: True when the action is allowed but needs user
            confirmation first (DEFAULT mode for write operations).
    """

    allowed: bool
    reason: str = ""
    requires_confirmation: bool = False

File: crates/ap-core/src/models/permission.rs
//! Permission models: `PermissionMode`, `PermissionConfig`, `PathRule`, `PermissionDecision`.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum PermissionMode {
    #[default]
    Default,
    Plan,
    FullAuto,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "kebab-case")]
pub enum PathAccess {
    #[default]
    Read,
    Write,
    ReadWrite,
    Deny,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PathRule {
    pub pattern: String,
    #[serde(default)]
    pub access: PathAccess,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct PermissionConfig {
    #[serde(default)]
    pub mode: PermissionMode,
    #[serde(default)]
    pub allowed_tools: Vec<String>,
    #[serde(default)]
    pub denied_tools: Vec<String>,
    #[serde(default)]
    pub path_rules: Vec<PathRule>,
    #[serde(default)]
    pub denied_commands: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PermissionDecision {
    pub allowed: bool,
    #[serde(default)]
    pub reason: String,
    #[serde(default)]
    pub requires_confirmation: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_permission_config() {
        let yaml = r#"
mode: default
allowed_tools: [file_read, grep]
denied_tools: [bash]
path_rules:
  - pattern: "*.docx"
    access: read-write
  - pattern: "*.env"
    access: deny
"#;
        let config: PermissionConfig = serde_yml::from_str(yaml).unwrap();
        assert_eq!(config.mode, PermissionMode::Default);
        assert_eq!(config.allowed_tools.len(), 2);
        assert_eq!(config.denied_tools, vec!["bash"]);
        assert_eq!(config.path_rules.len(), 2);
        assert_eq!(config.path_rules[1].access, PathAccess::Deny);
    }
}
